File: 213-house-robber-ii/house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        return Math.max(nums[0], 
                        Math.max(helper(nums, 0, nums.length - 1), 
                                 helper(nums, 1, nums.length)));
    }

    private static int helper(int[] nums, int startIdx, int endIdx) {
        int rob = 0;
        int noRob = 0;
        for (int i = startIdx; i < endIdx; i++) {
            int temp = rob;
            rob = noRob + nums[i];
            noRob = Math.max(temp, noRob);
        }
        return Math.max(rob, noRob);
    }
}

File: 213-house-robber-ii/house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        def dp(i, j):
            rob, no_rob = 0, 0
            for k in range(i, j):
                rob, no_rob = no_rob + nums[k], max(rob, no_rob)
            return max(rob, no_rob)

        return max(nums[0], dp(0, len(nums) - 1), dp(1, len(nums)))